src/multivariate/generaldiscretenonparametric.jl

```diff
@@ -0,0 +1,79 @@
+struct GeneralDiscreteNonParametric{VF,T,P <: Real,Ts <: AbstractVector{T},Ps <: AbstractVector{P},} <: Distribution{VF,Discrete}
+    support::Ts
+    p::Ps
+
+    function GeneralDiscreteNonParametric{VF,T,P,Ts,Ps}(
+        support::Ts,
+        p::Ps;
+        check_args=true,
+    ) where {VF,T,P <: Real,Ts <: AbstractVector{T},Ps <: AbstractVector{P}}
+        if check_args
+            length(support) == length(p) ||
+                error("length of `support` and `p` must be equal")
+            isprobvec(p) || error("`p` must be a probability vector")
+            allunique(support) || error("`support` must contain only unique values")
+        end
+        new{VF,T,P,Ts,Ps}(support, p)
+    end
+end
+
+function rand(rng::AbstractRNG, d::GeneralDiscreteNonParametric)
+    x = support(d)
+    p = probs(d)
+    n = length(p)
+    draw = rand(rng, float(eltype(p)))
+    cp = p[1]
+    i = 1
+    while cp <= draw && i < n
+        @inbounds cp += p[i +=1]
+    end
+    return x[i]
+end
+
+"""
+    support(d::MvDiscreteNonParametric)
+Get a sorted AbstractVector defining the support of `d`.
+"""
+support(d::GeneralDiscreteNonParametric) = d.support
+
+"""
+    probs(d::MvDiscreteNonParametric)
+Get the vector of probabilities associated with the support of `d`.
+"""
+probs(d::GeneralDiscreteNonParametric) = d.p
+
+
+Base.length(d::GeneralDiscreteNonParametric) = length(first(d.support))
+
+function _rand!(
+    rng::AbstractRNG,
+    d::GeneralDiscreteNonParametric,
+    x::AbstractVector{T},
+    ) where {T<:Real}
+
+    length(x) == length(d) || throw(DimensionMismatch("Invalid argument dimension."))
+    s = d.support
+    p = d.p
+
+    n = length(p)
+    draw = Base.rand(rng, float(eltype(p)))
+    cp = p[1]
+    i = 1
+    while cp <= draw && i < n
+        @inbounds cp += p[i+=1]
+    end
+    copyto!(x, s[i])
+    return x
+end
+
+
+function _logpdf(d::GeneralDiscreteNonParametric, x::AbstractVector{T}) where {T<:Real}
+    s = support(d)
+    p = probs(d)
+    for i = 1:length(p)
+        if s[i] == x
+            return log(p[i])
+        end
+    end
```

Comment:
In general, this could be optimized with `searchsortedfirst` if the support would be supported. At least for `ArrayLikeVariate` distributions sorting is clearly defined and works well, so maybe sort the support at least in these cases?
```julia
julia> sort([[0.0, 1.0, 0.0], [1.0, 0.0, 0.0], [0.5, 0.0, 0.5]])
3-element Vector{Vector{Float64}}:
 [0.0, 1.0, 0.0]
 [0.5, 0.0, 0.5]
 [1.0, 0.0, 0.0]

julia> searchsortedfirst(ans, [0.5, 0.0, 0.5])
```

Reply:
I actually have been implementing an algorithm that does this more efficiently. But I only got the 2D case working. My idea is to submit a future PR improving this.

Reply:
Why do we need a custom algorithm? Can't we just use `searchsortedfirst`? `sort` etc works with arrays just fine in Julia.

Reply:
I meant for computing the empirical cdf. Not exactly the sorting of the support.